src/main/java/com/magicvet/service/InputValidator.java

```diff
@@ -0,0 +1,32 @@
+package main.java.com.magicvet.service;
+
+import main.java.com.magicvet.Main;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class InputValidator {
+    public static String validateInputForPattern(String input, String patternString, String validInput) {
+        Pattern pattern = Pattern.compile(patternString);
+        Matcher matcher = pattern.matcher(prepareInputForMatch(input,patternString));
+        while (!matcher.matches()){
+            System.out.println("The input '" + input + "' is not valid! Valid input is: " + validInput);
+            System.out.print("Please try again: ");
+            input = Main.SCANNER.nextLine();
+            matcher = pattern.matcher(prepareInputForMatch(input,patternString));
+        }
+        return prepareInputForMatch(input,patternString);
+    }
+
+    private static String prepareInputForMatch (String input, String patternString) {
+        String patternStringWithoutRegexMetaChar = patternString.replaceAll(
+                "\\\\(d|D|s|S|t|n|w|W|b|B|Q|E|A|Z|z|G)","");
```

Comment:
The part of the code that removes metacharacters from the pattern may be considered redundant if we assume that the programmer has correctly defined the pattern.

Reply:
I would rather not rely on programmer in that case ;). Correctly define the pattern means to avoid the metacharacters in regex in this case. I don't think that it's convenient and always possible. I think that the method should be universal independent of the concrete pattern used. Later on the limitations of the method can be forgotten which would result in its incorrect behaviour for certain cases.
Imagine that we have decided to write the names of medical drugs and timings in uppercase. For ex.: "CALCIUM GLUCONATE-500 BID". (BID = two times a day). So the pattern for the prescription would be something like that: DRUG_PATTERN = "^[A-Z]+\\s*[A-Z]*-\\d+\\s+[A-Z]+$". One can see that DRUG_PATTERN.toUpperCase() is not equal to the DRUG_PATTERN due to the regex metacharacters (\\d,\\s). At the same time it's better to do input.toUpperCase() to facilitate the matching for the user input. That was the reason I've decided to get rid of the metacharacters in the patternString before comparing it to UpperCase/LowerCase.

Reply:
Got it. It's smart to make sure the method works for any pattern, even if it includes special characters. You're taking a good approach by removing those special characters from the pattern before comparing them. This way, it's easier to handle cases where patterns have different letter cases or special characters. Thanks for explaining why you're doing it this way.